### PR TITLE
Run PS Smoke tests with service principal

### PR DESCRIPTION
**Pull request by shahabhijeet**
This will allow to run our smoke tests to run without manual login.
This change enables to 
1) login in AzureRm using ServicePrincipal
2) login in ASM using publishsettings files
3) Stores all credentials in azure keyvault

Workflow
1) First time you run this tests, checks if certificate is installed
2) If not asks you to log in interactively so that Azure KeyVault can be accessed
3) Get's all the credentials/certs that are required to login
4) Logs in AzureRm and ASM
5) Once the test is done, removes subscription as well as any files that were downloaded.

This script has also the capability of create ServicePrincipal as well as new certificates if expired.
The logic to detect expired certificate is not implemented, but all the code that is required to
1) Create new certificates and store in Azure KeyVault
2) Create new SPN
3) Upload new PublishSettings file (if required)

### CODE REVIEW

File: tools/InstallationTests/Assert.ps1 [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿# ----------------------------------------------------------------------------------
```
Comment (markcowl) [inline tools/InstallationTests/Assert.ps1]: Please revert this change

File: tools/InstallationTests/AutoTestLogin.ps1 [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+    if($cert -eq $null)
+    {
+        Log-Info "Trying to find certificate in LocalMachine"
+        $cert = Get-ChildItem 'Cert:\LocalMachine\My' | Where-Object {$_.Subject -eq $Global:gLocalCertSubjectName}
+    }
+
+    return $cert
+}
+
+Function Get-ADAppForAutoTest()
```
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: This should be a cmdlet, so you can implement SHouldProcess/ShouldContinue

File: tools/InstallationTests/AutoTestLogin.ps1 [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+        $certPwd = Get-LocalCertificatePassword
+        $secCertPwd = ConvertTo-SecureString -String $certPwd -AsPlainText -Force
+        $pfxLocalFilePath = [System.IO.Path]::Combine($global:localPfxDirPath, $global:gPfxLocalFileName)
+        $localCert = New-Object System.Security.Cryptography.X509Certificates.X509Certificate($pfxLocalFilePath, $secCertPwd)        
+
+        $localCert.GetEffectiveDateString()
+        $localCert.GetExpirationDateString()
+        #>
+
+        $kvCertObj = Get-AzureKeyVaultCertificate -VaultName $global:gVaultName -Name $global:kvSecKey_PsAutoTestCertNameKey
```
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: This should be protected with ShouldProcess/ShouldContinue

File: tools/InstallationTests/AutoTestLogin.ps1 (hunk 2 above)
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: Also, wondering about the name, this should be something like Ensure, rather than get.  It should return the ADApp, so that you don't have to retrieve it again later

File: tools/InstallationTests/AutoTestLogin.ps1 [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+    $installedCert = Get-LocalCertificate
+    if($installedCert -eq $null)
+    {
+        throw [System.ApplicationException] "Unable to retrieve installed certificate for running automation test"
+    }
+    else
+    {
+        Log-Info "Successfully installed certificate to '$global:gLocalCertStore'"
+    }
+}
```
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: Remove the certificate (pfx file) when done

File: tools/InstallationTests/AutoTestLogin.ps1 [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+
+Function Login-AzureRMWithCertificate()
+{
+    $appId = $global:gPsAutoTestADAppId
+
+    $localCert = Get-LocalCertificate
+    if($localCert -eq $null)
+    {
+        #we could not find certificate installed locally, get it from KeyVault
+        Login-InteractivelyAndSelectTestSubscription
```
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: This should be protected by SHouldContinue, so it will just fail if run as part of a ccript.  This means the function should eb a cmdlet

File: tools/InstallationTests/AutoTestLogin.ps1 [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+    $global:gLoggedInCtx = Select-AzureRmSubscription -SubscriptionId $global:gPsAutoTestSubscriptionId
+
+    return $global:gLoggedInCtx
+}
+
+Function Login-Azure([bool]$deleteLocalCertificate=$false)
+{
+    Remove-AllSubscriptions
+
+    Delete-LocalCertificate $deleteLocalCertificate
```
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: This is kind of odd, why not use if ($deleteLoclCertificate)?

File: tools/InstallationTests/AutoTestLogin.ps1 [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+
+    Login-AzureRMWithCertificate
+    Select-AzureRmSubscription -SubscriptionId $global:gPsAutoTestSubscriptionId -TenantId $global:gTenantId
+
+    Login-AzureWithCertificate
+    Select-AzureSubscription -SubscriptionId $global:gPsAutoTestSubscriptionId -Current
+}
+
+<#
+Function Test-CleanUp()
```
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: Remove commented code

File: tools/InstallationTests/AutoTestLogin.ps1 [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+        $certCollection2 = New-Object System.Security.Cryptography.X509Certificates.X509Certificate2Collection
+        $certCollection2.Import($kvSecretBytes, $null, [System.Security.Cryptography.X509Certificates.X509KeyStorageFlags]::Exportable)
+        
+        #Save Pfx locally
+        $pwd = Get-LocalCertificatePassword
+        $certPwdProtectedInBytes = $certCollection2.Export([System.Security.Cryptography.X509Certificates.X509ContentType]::Pkcs12, $pwd)
+        $pfxLocalFilePath = [System.IO.Path]::Combine($global:localPfxDirPath, $global:gPfxLocalFileName)
+        [System.IO.File]::WriteAllBytes($pfxLocalFilePath, $certPwdProtectedInBytes)
+
+        #We are not going the .cer route for now, rather use publishSettings file
```
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: Remove commented code

File: tools/InstallationTests/AutoTestLogin.ps1 [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+# limitations under the License.
+# ----------------------------------------------------------------------------------
+
+This script will allow to login as long as you have required certificate installed on your machine
+
+WorkFlow# 1
+===========
+1) Will Check if the certificate is installed
+2) If not will ask you to login (one time) to access KeyVault
+3) Download Certificate, install certificate on the machine (will prompt to set certificate password)
```
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: Any prompts should be ShouldContinue prompts, so that they will fail when this is being run as part of script execution.  The idea is to allow extra bootstrapping as a convenience to manual users, but allowing automation to fail rapidly if the machine is not properly set up.

Also, just to be clear.  A properly set up machine should require no downloads from KeyVault, I think

File: tools/InstallationTests/AutoTestLogin.ps1 [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+    {
+        $certPath = $cert.PSPath
+        Log-Info "Deleting local certificate $certPath"
+        Remove-Item $cert.PSPath
+    }
+}
+
+Function CleanUpSPN()
+{
+    $psAutoADApp = Get-AzureRmADApplication -DisplayNameStartWith $global:gPsAutoTestADAppUsingSPNKey
```
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: Why do we remove the application?

File: tools/InstallationTests/AutoTestLogin.ps1 [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+}
+
+Function Get-KVSecretValue([string] $kvSecretName)
+{
+    return Get-AzureKeyVaultSecret -VaultName $global:gVaultName -Name $kvSecretName
+}
+
+Function Delete-TestCertificateFromKv([string] $vaultName, [string] $certName)
+{
+    Remove-AzureKeyVaultCertificate -VaultName $vaultName -Name $certName
```
Comment (markcowl) [inline tools/InstallationTests/AutoTestLogin.ps1]: ??

File: tools/InstallationTests/Common.ps1 [hunk 12]
```diff
@@ -193,14 +193,28 @@ function Test-Setup
 #############################
 function Test-Cleanup
```
Comment (markcowl) [inline tools/InstallationTests/Common.ps1]: Please ensure that these are the latest versions from the PowerShell share.  The files here are quite out of sync with the scripts we have been executing.

File: tools/InstallationTests/AutoTestLogin.ps1 (hunk 9 above)
Comment (shahabhijeet) [inline tools/InstallationTests/AutoTestLogin.ps1]: ShouldContinue will be implemented when we decide to make the entire login script to a module. Currently added a switch that will allow to fail fast on CI machine.

File: tools/InstallationTests/AutoTestLogin.ps1 (hunk 2 above)
Comment (shahabhijeet) [inline tools/InstallationTests/AutoTestLogin.ps1]: Extracted one time machine setup function to a separate psm1 file which will eventually a script module.

File: tools/InstallationTests/AutoTestLogin.ps1 (hunk 4 above)
Comment (shahabhijeet) [inline tools/InstallationTests/AutoTestLogin.ps1]: Added Try/Finally to remove all downloaded resources

File: tools/InstallationTests/AutoTestLogin.ps1 (hunk 6 above)
Comment (shahabhijeet) [inline tools/InstallationTests/AutoTestLogin.ps1]: Fixed the logic

File: tools/InstallationTests/AutoTestLogin.ps1 (hunk 10 above)
Comment (shahabhijeet) [inline tools/InstallationTests/AutoTestLogin.ps1]: Moved code to a separate file.

File: tools/InstallationTests/Common.ps1 [hunk 13]
```diff
@@ -193,14 +193,33 @@ function Test-Setup
...
+     #$global:VerbosePreference = $global:oldVerbosePreference
+     #$global:WarningPreference = $global:oldWarningPreference
+     #$global:WhatIfPreference = $global:oldWhatIfPreference
+
+    Delete-DownloadedCertAndPubSetting
+    
+    Remove-AllSubscriptions
+}
+
+Function Delete-DownloadedCertAndPubSetting
```
Comment (markcowl) [inline tools/InstallationTests/Common.ps1]: Seems like this belongs in the same file as the code that downloads the cert and publishsettings

Comment (shahabhijeet) [inline tools/InstallationTests/Common.ps1]: Moved to the same that creates these files.
